dts/bindings/dma/st,stm32-dma-v3.yaml

```
@@ -0,0 +1,106 @@
+  - bit 15: Reserved
+  - bits 16-17: Priority Level
+            0x0: STM32_DMA_PRIORITY_LOW: Low
+            0x1: STM32_DMA_PRIORITY_MEDIUM: Medium
+            0x2: STM32_DMA_PRIORITY_HIGH: High
+            0x3: STM32_DMA_PRIORITY_VERY_HIGH: Very High
+  - bit 18: Source Burst Increment
+            0x0: STM32_DMA_DEST_BURST_NO_INC: no burst increment
+            0x1: STM32_DMA_DEST_BURST_INC: burst increment
+  - bit 19: Source Allocated Port
+            0x0: STM32_DMA_SRC_PORT_0: Port 0
+            0x1: STM32_DMA_SRC_PORT_1: Port 1
+  - bit 20: Destination Burst Increment
+            0x0: STM32_DMA_DEST_BURST_NO_INC: no burst increment
+            0x1: STM32_DMA_DEST_BURST_INC: burst increment
+  - bit 21: Destination Allocated Port
+            0x0: STM32_DMA_DEST_PORT_0: Port 0
+            0x1: STM32_DMA_DEST_PORT_1: Port 1
+  - bit 22: Hardware Request Type
+            0x0: STM32_DMA_HW_BURST_REQUEST_MODE: Burst
+            0x1: STM32_DMA_HW_BLOCK_REQUEST_MODE: Block
+  - bit 23: Control Mode
+            0x0: STM32_DMA_CTRL_MODE_DMA: DMA controller control mode
+            0x1: STM32_DMA_CTRL_MODE_PERIPH: Peripheral control mode
+  - bits 24-25: Transfer Complete Event Mode
+            0x0: STM32_DMA_TC_EVENT_BLOCK: At block level
+            0x2: STM32_DMA_TC_EVENT_LLI: At Linked List Item level
+            0x3: STM32_DMA_TC_EVENT_CHANNEL: At channel level
+  - bit 26: Packing/Unpacking Mode
+            0x0: STM32_DMA_ALLOW_PACKING: Allow
+            0x1: STM32_DMA_PREVENT_PACKING: Prevent
+  - bit 27: Additional Transfers
+            0x0: STM32_DMA_ALLOW_ADDITIONAL_TRANSFERS: Allow
+            0x1: STM32_DMA_PREVENT_ADDITIONAL_TRANSFERS: Prevent
+
+  For the client part, example for stm32mp257f-ev1 on the CM33 using the HPDMA3
+  instance:
+    rx using channel 0 with request 17 for uart5_rx
+    tx using channel 1 with request 18 for uart5_tx
+    uart5 {
+      dmas = <&hpdma3 0 18 (STM32_DMA_PERIPH_TX | STM32_DMA_PRIORITY_HIGH)>,
```

typically these sorts of dma cells have solely a node reference and channel identifier, maybe a handshake peripheral number. The choice to add bit field options here for linked list, cyclic, or various other modes of operation is interesting but I think gives the illusion that perhaps changing these fields will actually change the way the DMA is used at a peripheral driver level.

Is that actually the intent here? To allow say, a uart to work in linked list or cyclic mode, with 16, 32 or 64bit data words? That doesn't seem logical to me.

---

That's my point, the uart driver is going to configure the DMA for how it wishes to use it, having these fields here is confusing and would imply to someone configuring their device that these are options that can be changed for say... how the uart driver may use the DMA, but is that really the case? Can I change say how the st uart driver uses DMA to use the linked list descriptors or cyclic buffer mode? That would, I think, be a very different looking driver.

---

So what you re asking for is either get the uart driver to actually be able to use Linked Lists,  (I don t see the use of having 32bits transfers as the hardware will only transfer 8 anyways), and remove all the unused parameters from the dts binding ? 

Then again how would you have a custom configuration for Hardware transfer requests ?

---

I'm saying don't provide bits in DTS defines that don't make sense from a user configuration perspective, as that's what it is. It's user configurable. If a driver wishes to set word sizing, linked list or cyclic, what the watermark would be, etc etc for a DMA transfer that's entirely up to the *driver* using the peripheral to do this. Otherwise what I'd expect is that a poor user comes along, sees that ST supports linked list transfers or cyclic transfers with their DMA and any peripheral, only to confusingly either break the peripheral driver or be silently ignored. Either of those outcomes is a poor user experience for someone setting up their devicetree.

---

> That's my point, the uart driver is going to configure the DMA for how it wishes to use it, having these fields here is confusing and would imply to someone configuring their device that these are options that can be changed for say... how the uart driver may use the DMA, but is that really the case? Can I change say how the st uart driver uses DMA to use the linked list descriptors or cyclic buffer mode? That would, I think, be a very different looking driver.

It seems to be the case, actually: DMA configuration `cyclic` depends on DT property
https://github.com/zephyrproject-rtos/zephyr/blob/9135918249b36dbbd675db8c4f94173c147438f3/drivers/serial/uart_stm32.c#L2377-L2378

and the driver checks whether it's enabled or not during its operation:
https://github.com/zephyrproject-rtos/zephyr/blob/9135918249b36dbbd675db8c4f94173c147438f3/drivers/serial/uart_stm32.c#L1162-L1173

Whether all drivers support all use cases is another question but there are at least *some* drivers that consume *some* options to affect their behavior.

(+ I don't remember the exact DMA API, but as far as I understand, whether a DMA transfer is done with a "direct" programming or a linked list should be an implementation detail of the DMA driver?)

---

> whether a DMA transfer is done with a "direct" programming or a linked list should be an implementation detail of the DMA driver? 

It seems that it is up to the driver to use one of the two, but that is depending on the usage/configuration. 
The DMA struct from Zephyr enables the use of chained blocks, so one way is saying if you have more than one block to configure linked lists, one can also juste configure a said channel for a transfer and call the `reload` function to do another transfer in direct programming with changing only the addresses, having the same effect with only direct programming. 

So I d say the usage of linked lists or direct programmign depends on how the user side uses the DMA API and its structure.

In my case I tried exposing all the ways depending on the available tests. Si linked lists are used with cyclic and scatter gather tests as they need multiple transfers. and it is better to do so fully with DMA rather than call the CPU at each end of transfer. 

Either way, some of the symbols I added in the yaml are in fact not consumable so far so I guess I ll just remove them for next batch and add them only when that s the case. :D

---

@teburd PTAL :p